File: src/main/java/com/github/njuro/jard/base/BaseFacade.java
package com.github.njuro.jard.base;

import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * Base facade providing suitable implementation of {@link BaseMapper} and delegate methods for
 * shorter code.
 */
@Component
public abstract class BaseFacade<ENTITY extends BaseEntity, DTO extends BaseDto> {

  private BaseRepository<ENTITY> repository;

  protected BaseMapper<ENTITY, DTO> mapper;

  protected DTO toDto(ENTITY entity) {
    return mapper.toDto(entity);
  }

  protected List<DTO> toDtoList(List<ENTITY> entityList) {
    return mapper.toDtoList(entityList);
  }

  protected ENTITY toEntity(DTO dto) {
    if (dto.getId() != null) {
      ENTITY existing = repository.findById(dto.getId()).orElseThrow();
      return mapper.toExistingEntity(dto, existing);
    }

    return mapper.toEntity(dto);
  }

  protected List<ENTITY> toEntityList(List<DTO> dtoList) {
    return mapper.toEntityList(dtoList);
  }

  @Autowired
  private void setMapper(BaseMapper<ENTITY, DTO> mapper) {
    this.mapper = mapper;
  }

  @SuppressWarnings("SpringJavaInjectionPointsAutowiringInspection")
  @Autowired
  private void setRepository(BaseRepository<ENTITY> repository) {
    this.repository = repository;
  }
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/base/BaseFacade.kt
package com.github.njuro.jard.rewrite.base

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

/**
 * Base facade providing suitable implementation of [BaseMapper] and delegate methods for
 * shorter code.
 */
@Component
abstract class BaseFacade<ENTITY : BaseEntity, DTO : BaseDto> {

    @Autowired
    private lateinit var repository: BaseRepository<ENTITY>

    @Autowired
    protected lateinit var mapper: BaseMapper<ENTITY, DTO>

    protected fun toDto(entity: ENTITY): DTO = mapper.toDto(entity)

    protected fun toDtoList(entityList: List<ENTITY>): List<DTO> = mapper.toDtoList(entityList)

    protected fun toEntity(dto: DTO): ENTITY {
        if (dto.id != null) {
            val existing = repository.findById(dto.id).orElseThrow()
            return mapper.toExistingEntity(dto, existing)
        }
        return mapper.toEntity(dto)
    }

    protected fun toEntityList(dtoList: List<DTO>): List<ENTITY> = mapper.toEntityList(dtoList)
}
